File: CPP/2-Structurals/Bridge/Bridge.cpp
#include <iostream>
#include <string>

// Implemented interface.
class WidgetAbstractInterface
{
    public:
        virtual ~WidgetAbstractInterface() {}
        virtual void createWidget() = 0;
};

// Interface for internal implementation that Bridge uses.
class ImplementationInterface
{
    public:
        virtual ~ImplementationInterface() {}
        virtual void setupWidget() = 0;
};

// The Bridge
class Bridge : public WidgetAbstractInterface
{
    protected:
        ImplementationInterface *_implementation;

    public:
        Bridge(ImplementationInterface *implementation)
        {
            _implementation = implementation;
        }
};

// Different special cases of the interface.
class Button : public Bridge
{
    public:
        Button(ImplementationInterface *implementation)
          : Bridge(implementation)
        {}

        void createWidget()
        {
            std::cout << "Button on ";
            _implementation->setupWidget();
        }
};

class TextField : public Bridge
{
    public:
        TextField(ImplementationInterface *implementation)
          : Bridge(implementation)
        {}

        void createWidget()
        {
            std::cout << "TextField on ";
            _implementation->setupWidget();
        }
};

// Different background implementations
class Windows : public ImplementationInterface
{
    public:
        void setupWidget()
        {
            std::cout << "Windows..." << std::endl;
        }
};

class Linux : public ImplementationInterface
{
    public:
        void setupWidget()
        {
            std::cout << "Linux..." << std::endl;
        }
};

int main()
{
    WidgetAbstractInterface *widget = 0;
    ImplementationInterface *osWindows = new Windows;
    ImplementationInterface *osLinux = new Linux;

    // First case
    widget = new Button(osWindows);
    widget->createWidget();
    // ... do something
    delete widget;

    widget = new Button(osLinux);
    widget->createWidget();
    // ... do something
    delete widget;

    // Second case
    widget = new TextField(osWindows);
    widget->createWidget();
    // ... do something
    delete widget;

    widget = new TextField(osLinux);
    widget->createWidget();
    // ... do something
    delete widget;

    delete osWindows;
    delete osLinux;

    return 0;
};

//-----------------------------------------------------------
/*
Output:

Button on Windows...
Button on Linux...
TextField on Windows...
TextField on Linux...

*/

File: Python/2-Structurals/Bridge/Bridge.py
#!/usr/bin/env python3

from abc import ABC, abstractmethod


class WidgetAbstractInterface(ABC):
    """
    Abstract class for the widget's API
    """
    def __init__(self):
        pass

    @abstractmethod
    def create_widget(self):
        pass


class ImplementationInterface(ABC):
    """
    Abstract class for the widget's implementation
    """
    def __init__(self):
        pass

    @abstractmethod
    def setup_widget(self):
        pass


class Bridge(WidgetAbstractInterface):
    """
    The bridge abstract class
    """
    def __init__(self, implementation):
        WidgetAbstractInterface.__init__(self)
        self.implementation_interface = implementation

    def create_widget(self):
        pass


class Button(Bridge):
    """
    Button interface over different implementations
    """
    def create_widget(self):
        print("Button on ")
        self.implementation_interface.setup_widget()


class TextField(Bridge):
    """
    Text field interface over different implementations
    """
    def create_widget(self):
        print("TextField on ")
        self.implementation_interface.setup_widget()


class Windows(ImplementationInterface):
    """
    Widget implementation on Windows
    """
    def setup_widget(self):
        print("Windows...\n")


class Linux(ImplementationInterface):
    """
    Widget implementation on Linux
    """
    def setup_widget(self):
        print("Linux...\n")


def main():
    os_windows = Windows()
    os_linux = Linux()

    # Create a button for the Windows OS
    win_button = Button(os_windows)
    win_button.create_widget()

    # Create a button for the Linux OS
    lin_button = Button(os_linux)
    lin_button.create_widget()

    # Create a TextField for the Windows OS
    win_button = TextField(os_windows)
    win_button.create_widget()

    # Create a TextField for the Linux OS
    lin_button = TextField(os_linux)
    lin_button.create_widget()


if __name__ == "__main__":
    main()


"""
Output:

Button on 
Windows...

Button on 
Linux...

TextField on 
Windows...

TextField on 
Linux...

"""
